File: 0229-Majority Element II/0229-Majority Element II.cpp
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        vector<int> result;
        int n = nums.size();
        if (n == 0) {
            return result;
        }
        int num1 = 1, num2 = 2;
        int count1 = 0, count2 = 0;
        for (int num : nums) {
            if (num == num1) {
                ++count1;
            }
            else if (num == num2) {
                ++count2;
            }
            else if (count1 == 0) {
                num1 = num;
                count1 = 1;
            }
            else if (count2 == 0) {
                num2 = num;
                count2 = 1;
            }
            else {
                --count1;
                --count2;
            }
        }
        
        count1 = count2 = 0;
        for (int num : nums) {
            if (num == num1) {
                ++count1;
            }
            if (num == num2) {
                ++count2;
            }
        }
        
        if (count1 > n / 3) {
            result.push_back(num1);
        }
        
        if (count2 > n / 3) {
            result.push_back(num2);
        }
        
        return result;
    }
};

File: 0201-0300/0229-Majority Element II/0229-Majority Element II.py
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        count1 = count2 = 0
        num1 = 1
        num2 = 2
        for num in nums:
            if num == num1:
                count1 += 1
            elif num == num2:
                count2 += 1
            elif count1 == 0:
                num1 = num
                count1 = 1
            elif count2 == 0:
                num2 = num
                count2 = 1
            else:
                count1 -= 1
                count2 -= 1
        
        return [num for num in (num1, num2) if nums.count(num) > len(nums) // 3]
